### DISCUSSION TITLE
Desktop application's runtime environment

### DISCUSSION THREAD

**Discussion by eeroleppalehto** [eeroleppalehto/Jahtirekisteri#134] [Ideas]
Currently the code in the `PyQt` desktop application is run synchronously which freezes the application for too long when the application is fetching data from a remote server.

As such there is a choice to be had on how the application's code ought to be ran.

### Asynchronous

One solution for the UI to freeze up is to change the code to run asynchronously. Problem with running the code asynchronously is that PyQt5 doesn't allow that out of the box and requires external libraries for achieving that.

Example libraries:
[qt-async-threads](https://qt-async-threads.readthedocs.io/en/latest/)
[qasync](https://pypi.org/project/qasync/)

### Multithreading or Worker thread

Other solution is to use `Qt`'s the built in `QThread` class to run the code that fetches data on a separate thread. This would resolve the issue of the UI freezing up.

Tutorials on `QThread`:

[pythonguis.com](https://www.pythonguis.com/tutorials/multithreading-pyqt-applications-qthreadpool/)
[realpython.com](https://realpython.com/python-pyqt-qthread/)

### Synchronous

Last option would just keep the code as it is and try lower the amount calls to the server to limit time that UI is frozen. Although this can also be done for the other two options to further optimize them

I would like thoughts on these so one of these options can be chosen and further developed on.

**Comment by VerneriLahteenoja** [comment]
`QThread` Seems like a solid option. Considerably less development time compared to asynchronous, although could still face a bit of issues with database calls once the data is fetches from a remote server in large quantities.
I'll create a separate branch from dev to test QThread out with the existing structures.
